### DISCUSSION TITLE
Trust custom/self-signed CA

### DISCUSSION THREAD

**Discussion by presianbg** [Security-Onion-Solutions/securityonion#3392] [Unsupported Versions] [answered]
Hi,

We are having sensors in network segments where SSL\TLS inspection (decryption by middle-box) is done. 
Because the Zeek does not have our internal trusted CA we have a lot of notices:

```
SSL::Invalid_Server_Cert - SSL certificate validation failed with (unable to get local issuer certificate)
```

After some googling and reading the docs I found we can `redef` the `SSL::root_certs` list.

Looks like Bro/Zeek require its own cert format and I was only able to find this script for converting the cert:

```bash
#! /usr/bin/perl
use Encode;
use Crypt::OpenSSL::X509;

while (<>) {
  if (/BEGIN /) {
    $p="";
    $inside=1;
  }
  $p .= $_ if $inside;
  if (/END /) {
    $c = Crypt::OpenSSL::X509->new_from_string($p);
    @s = map { $_->as_string } reverse @{$c->subject_name->entries};
    s/[\\,]/\\$&/g for @s;
    $s = join ",", @s;
    $s = encode("UTF-8", $s);
    $s =~ s/[\200-\377]/sprintf("\\%X",ord$&)/ge;
    $x = join "", map { "\\x" . uc$_ } unpack("(H2)*", $c->as_string(Crypt::OpenSSL::X509::FORMAT_ASN1));
    print "\t[\"$s\"] = \"$x\",\n"; $inside = 0;
  }
}
```

So I created a local policy in:
`/opt/so/saltstack/local/salt/zeek/policy/custom/trust_custom_ca`
ls -la:
```
ls -la
total 12
drwxr-xr-x. 2 socore socore   55 Mar 11 09:14 .
drwxr-xr-x. 4 socore socore   53 Mar 11 08:34 ..
-rw-r--r--. 1 socore socore   24 Mar 11 08:40 __load__.zeek
-rw-r--r--. 1 socore socore 4801 Mar 11 09:14 trust_custom_ca.zeek
```
`__load__.zeek:`
```
@load ./trust_custom_ca
```
`trust_custom_ca.zeek`:

```
@load base/protocols/ssl

redef SSL::root_certs += {
	["emailAddress=support@blabla.local,CN=Blabla SSL CA_C24077D87BQMV67,OU=NSG,O=BlaBla,ST=BlaBla,C=GB"] = "\x30\x82\x04\x89\x30\x82\x03\x71\xA0\x03\x02\x01\x02\x02\x01.....<snip>"
};
```
(also tried without having `@load base/protocols/ssl`)

And adding `- custom/trust_custom_ca` in `/opt/so/saltstack/local/pillar/zeek/init.sls`:

```
<snip>
      - securityonion/communityid
      - securityonion/file-extraction
      - custom/filter_dns_noise
      **- custom/trust_custom_ca**
    '@load-sigs':
      - frameworks/signatures/detect-windows-shells
<snip>
```

Doing `sudo salt \* state.highstate` results in:

`ERROR: Minions returned with non-zero exit code` and zeek stops working.

This is distributed environment (latest version 2.3.30)  with single manager node + 2 search nodes + 3 sensors if that matters. 

Would you be so kind to help me finding why this is not working ?

Kind regards,
Presian

**Comment by weslambert** [comment]
You might try something like the following to diagnose the issue with Zeek:

`docker exec -it so-zeek /opt/zeek/bin/zeekctl check`

**Reply by presianbg** [reply]
Thank you. I will try that.
I see another discussion with similar topic:
https://github.com/Security-Onion-Solutions/securityonion/discussions/1877

**Reply by presianbg** [reply]
I found useful scrip, which I believe would be useful for others and maybe worth including in the How-to (Docs), because SSL\TLS inspection is getting more popular:

https://gist.github.com/srozb/d1e5770e4fe9f0f60a4fca4a57270da6

My mistake was that I was not loading the proper script/module to redefine.

**Reply by weslambert** [reply]
If you have resolved this, please post the answer in a separate comment and mark it as the answer.  Thanks!

**Comment by presianbg** [comment, marked as answer]
I managed to add a custom CA to the zeek config and now SSL::Invalid_Server_Cert notice messages dropped drastically. 
All I was missing from the original post was `module SSL;` directive. 
This script found in gists also helps for automatic the process of adding new CAs:

```bash
#!/bin/bash

INFORM="DER"

echo @load base/protocols/ssl
echo
echo module SSL\;
echo redef root_certs += {

for f in *.cer
do
    CN=$(openssl x509 -inform $INFORM -in $f -noout -subject | cut -c 11- | sed 's/\//,/g')
    CERTDATA=$(openssl x509 -inform $INFORM -in $f -outform DER | hexdump -v -e '"\\""x" /1 "%02X"')
    echo [\"$CN\"] = \"$CERTDATA\",
done

echo }\;
```

**Reply by TOoSmOotH** [reply]
This is pretty cool thanks for posting!

**Reply by corycandia** [reply]
@presianbg,

Is your marked answer the only required script?

Where did the final script go?

Sorry, I'm clearly not as well versed in the *nix world, but we also do ssl inspection with an inline sophos utm and a trusted signing cert that I'd love to be able to monitor with SO.  This is the only thing I feel I'm missing with the security onion solution out of the box.

God bless SO!

**Reply by presianbg** [reply]
Hi @corycandia,

This are the steps / scripts you need:
1. Place the following in `/opt/so/saltstack/local/salt/zeek/policy/custom/trust_custom_ca/__load__.zeek`:
```
@load ./trust_custom_ca
```
and  `/opt/so/saltstack/local/salt/zeek/policy/custom/trust_custom_ca/trust_custom_ca.zeek`
```
@load base/protocols/ssl

module SSL;
redef SSL::root_certs += {
	["C = GB, ST = Oxfordshire, O = Sophos, OU = NSG, CN = Sophos SSL CA_XXXX, emailAddress = support@sophos.com"] = "\x30\x82\x04\x89\x30\x82\x03\x71\xA0\x03\x02\x01\x02\x02\x01\x01\x30\x0D\x06\x.....
[snip]
BC\x27\xAD\xC8",
};
```

You need to convert the CA you  exported from the Sophos middlebox to hex with [this](https://github.com/Security-Onion-Solutions/securityonion/discussions/3392#discussioncomment-493651) script.

2. In `/opt/so/saltstack/local/pillar/zeek/init.sls` add - `- custom/trust_custom_ca` in the end of :

```
  local:
    '@load':
```

Be aware that not all of the Zeek notices -  `SSL::Invalid_Server_Cert` will disappear, but their count will drop significantly. 

Cheers,
PY
